# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_id_text = [
        [source_vocab_to_int[word] for word in sentence.split()] for sentence in source_text.split('\n')]
    target_id_text = [
        [target_vocab_to_int[word] for word in sentence.split()] + [target_vocab_to_int['<EOS>']]
        for sentence in target_text.split('\n')]

    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [31]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [32]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [33]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input_text     = tf.placeholder(dtype=tf.int32, shape=[None,None], name="input")
    target_text    = tf.placeholder(dtype=tf.int32, shape=[None,None], name="target")
    learning_rate  = tf.placeholder(dtype=tf.float32, shape=None, name="learning_rate")
    keep_prob      = tf.placeholder(dtype=tf.float32, shape=None, name="keep_prob")
    target_sequence_length = tf.placeholder(dtype=tf.int32, shape=[None], name="target_sequence_length")
    max_target_len = tf.reduce_max(target_sequence_length, axis=0, name="max_target_len")
    source_sequence_length = tf.placeholder(dtype=tf.int32, shape=[None], name="source_sequence_length")
    return (input_text, target_text, learning_rate, keep_prob,
           target_sequence_length, max_target_len, source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [34]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, begin=[0,0], end=[batch_size,-1], strides=[1,1]) 
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)

    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [35]:
from imp import reload
reload(tests)

<module 'problem_unittests' from '/home/carnd/Krbn/deep-learning/language-translation/problem_unittests.py'>

In [36]:
# LSTM or GRU cells
# Here I create a function that creates a single GRU or LSTM cell
#   w/ optional dropout
def single_RNN_cell(rnn_size, keep_prob=None, cell_type="LSTM"):
    """
    Create single RNN cell (LSTM or GRU) w/ optional dropout
    :param rnn_size: RNN Size
    :param keep_prob: Dropout keep probability
    :param cell_type: "LSTM" (default) or "GRU" (not case sensitive)
    """
    cell_type = cell_type.lower()[0]
    if cell_type == "l":  
        # LSTM CELL
        single_cell = tf.contrib.rnn.LSTMCell(
            rnn_size,
            initializer = tf.random_uniform_initializer(-0.1, 0.1, seed=2)
        )
    else:          
        #  GRU CELL
        single_cell = tf.contrib.rnn.GRUCell(
            rnn_size
        )
    if keep_prob != None:
        single_cell = tf.contrib.rnn.DropoutWrapper(
            single_cell, 
            output_keep_prob = keep_prob
        )
    return single_cell
    

In [37]:
# Here I allow the user to create an encoding layer based on LSTM cells,
#   GRU cells, or both.  To use both, the user need to pass a list to
#   cell_type that defines a pattern, e.g., if num_layers=10, cell_type=["l","g"]
#   will produce an encoder cell of the form ["l", "g"]*5.
def encoding_layer(
    rnn_inputs, 
    rnn_size, 
    num_layers, 
    keep_prob, 
    source_sequence_length, 
    source_vocab_size, 
    encoding_embedding_size, 
    cell_type = "LSTM"):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :param cell_type: "LSTM" (default), "GRU", or list of these (not case sensitive)
    :return: tuple (RNN output, RNN state)
    """
    #---------------------------------
    # TODO: Implement Function
    #---------------------------------
    
    # Embed the Encoder Input Sequence
    encoder_embedding = tf.contrib.layers.embed_sequence(
        rnn_inputs, 
        source_vocab_size, 
        encoding_embedding_size
    )
    
    # Encoder Cell is a Stacked (Multi-Layer) RNN Cell
    if str(type(cell_type)) == "<class 'str'>": 
        cell_type = [cell_type]
    if len(cell_type) == 1:
        # Homogeneous Layers
        cell_type = cell_type * num_layers
    else:
        # Heterogeneous Layers
        N = len(cell_type)
        M = int(np.ceil(num_layers/N))
        cell_type = (cell_type * M)[0:num_layers]
    encoder_cell = tf.contrib.rnn.MultiRNNCell(
        [single_RNN_cell(rnn_size, keep_prob=keep_prob, cell_type=cell_type[i]) 
         for i in range(num_layers)]
    ) 
    # 
    encoder_output, encoder_state = tf.nn.dynamic_rnn(
        encoder_cell, 
        encoder_embedding, 
        sequence_length=source_sequence_length, 
        dtype=tf.float32
    )
    return encoder_output, encoder_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [38]:
# NOTE:
#   Seems like some of these inputs are red herrings... For example,
#   keep_prob is applied at the single RNN cell level of the MultiRNN cell
#   that makes up the decoder (dec_cell), which is an object that is 
#   presumably pre-existing since it is also an input argument to this
#   function...  
def decoding_layer_train(
    encoder_state,            # Used in BasicDecoder
    dec_cell,                 # Used in BasicDecoder
    dec_embed_input,          # Used in TrainingHelper
    target_sequence_length,   # Used in TrainingHelper
    max_summary_length,       # Used in dynamic_decode
    output_layer,             # Used in BasicDecoder (though given parameter description is weird)
    keep_prob,                # NOT EXPLICITLY USED IN TrainingHelper, BasicDecoder, or dynamic_decode
    parallel_iterations=32):  # Used in dynamic_decode
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :param parallel_iterations: Argument passed to tf.while_loop
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    training_helper = tf.contrib.seq2seq.TrainingHelper(
        inputs          = dec_embed_input,
        sequence_length = target_sequence_length,
        time_major      = False,
        name            = "training_helper"
    ) # RETURNED sample_ids are the argmax of the RNN output logits.
    
    training_decoder = tf.contrib.seq2seq.BasicDecoder(
        cell          = dec_cell,
        helper        = training_helper,
        initial_state = encoder_state,
        output_layer  = output_layer
    ) # RETURNS:  (outputs, next_state, next_inputs, finished)
    
    training_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(
        decoder            = training_decoder,
        output_time_major  = False,
        impute_finished    = True,
        maximum_iterations = max_summary_length,
        parallel_iterations= parallel_iterations
    ) # RETURNS:  (final_outputs, final_state, final_sequence_lengths)
    
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [39]:
vocab_size = len(target_vocab_to_int)

# NOTE:
#   Again, it seems like some of these inputs are red herrings... 
#   keep_prob has the same issue as described in the note above. 
#   We also have vocab_size which is not explicitly used...
def decoding_layer_infer(
    encoder_state,               # Used in BasicDecoder
    dec_cell,                    # Used in BasicDecoder
    dec_embeddings,              # Used in GreedyEmbeddingHelper
    start_of_sequence_id,        # Used in tf.tile (start_tokens)
    end_of_sequence_id,          # Used in GreedyEmbeddingHelper
    max_target_sequence_length,  # Used in dynamic_decode
    vocab_size,  
    output_layer,                # Used in BasicDecoder
    batch_size,                  # Used in tf.tile (start_tokens)
    keep_prob,
    parallel_iterations=32):     # Used in dynamic_decode
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS ID
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :param parallel_iterations: Argument passed to tf.while_loop
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
       
    start_tokens = tf.tile(
        #tf.constant([target_vocab_to_int['<GO>']], dtype=tf.int32), 
        tf.constant([start_of_sequence_id], dtype=tf.int32), 
        [batch_size], 
        name='start_tokens'
    ) 
    # RETURNS new tensor by replicating its input tensor
    #   (the tf.constant above) batch_size times.
    
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(
        embedding    = dec_embeddings,
        start_tokens = start_tokens,
        #end_token    = target_vocab_to_int['<EOS>']
        end_token    = end_of_sequence_id
    ) 
    # A helper for use during inference. Uses the argmax of the output 
    #  (treated as logits) and passes the result through an embedding layer 
    #  to get the next input.

    inference_decoder = tf.contrib.seq2seq.BasicDecoder(
        cell          = dec_cell,
        helper        = inference_helper,
        initial_state = encoder_state,
        output_layer  = output_layer
    ) # RETURNS:  (outputs, next_state, next_inputs, finished)
        
    inference_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(
        decoder             = inference_decoder,
        impute_finished     = True,
        maximum_iterations  = max_target_sequence_length,
        parallel_iterations = parallel_iterations
    ) # RETURNS:  (final_outputs, final_state, final_sequence_lengths)
    
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [40]:
# NOTE:
#   I noticed in the seq2seq code that the decoder and encoder embeddings are
#   done differently...  
#     * For the encoder embedding: 
#         embedding1 = tf.contrib.layers.embed_sequence(inputs, vocab_size, embedding_size)
#     * For the decoder embedding:
#         w_emb = tf.Variable(tf.random_uniform([vocab_size, embedding_size]))
#         embedding2 = tf.nn.embedding_lookup(w_emb, inputs)
#   
#    The following code was used to compare the two embedding tensors:    
#        vocab_size = len(target_vocab_to_int)
#        inputs     = tf.placeholder(dtype=tf.int32, shape=[None,None], name="target")
#        embedding_size = 100
#        e1 = tf.contrib.layers.embed_sequence(inputs, vocab_size, embedding_size)
#        w_emb = tf.Variable(tf.random_uniform([vocab_size, embedding_size]))
#        e2 = tf.nn.embedding_lookup(w_emb, inputs)
#
#    Results
#        e1: <tf.Tensor 'EmbedSequence/embedding_lookup:0' shape=(?, ?, 100) dtype=float32>
#        e2: <tf.Tensor 'embedding_lookup:0' shape=(?, ?, 100) dtype=float32>
target_vocab_size = len(target_vocab_to_int)

def decoding_layer(
    dec_input,                   # Used to create embedding
    encoder_state,               # Used in decoder_layer_train
    target_sequence_length,      # Used in decoder_layer_train
    max_target_sequence_length,  # Used in decoder_layer_train
    rnn_size,                    # Used to create decoder cell
    num_layers,                  # Used to create decoder cell
    target_vocab_to_int,         # Used to get GO & EOS IDs
    target_vocab_size,           # Used to create output layer, inference layer
    batch_size,                  # Used in decoder_layer_infer
    keep_prob,                   # Used to create decoder cell
    decoding_embedding_size,     # Used to create embedding
    cell_type = "LSTM",          # Used to create decoder cell
    parallel_iterations=32):      
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param cell_type: "LSTM" (default), "GRU", or list of these (not case sensitive)
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    # Embed Target Sequences (Decoder Inputs)
    decoder_embedding = tf.Variable(
        tf.random_uniform([target_vocab_size, decoding_embedding_size])
    )
    dec_embed_input = tf.nn.embedding_lookup(
        decoder_embedding, 
        dec_input
    )
    
    # Construct the Decoder Cell as a Stacked (Multi-Layer) RNN cell
    #   -- project requests stacked LSTM cells, but I provided GRU and LSTM-GRU 
    #      hybrid options as well
    #   -- w/ optional dropout feature 
    if str(type(cell_type)) == "<class 'str'>": 
        cell_type = [cell_type]
    if len(cell_type) == 1:
        # Homogeneous Layers
        cell_type = cell_type * num_layers
    else:
        # Heterogeneous Layers
        N = len(cell_type)
        M = int(np.ceil(num_layers/N))
        cell_type = (cell_type * M)[0:num_layers]
    #-----------------------------------------------
    dec_cell = tf.contrib.rnn.MultiRNNCell(
        [single_RNN_cell(rnn_size, keep_prob=keep_prob, cell_type=cell_type[i]) 
         for i in range(num_layers)]
    ) 
    
    # Dense Output Layer:
    #   Translates the decoder's output into a choice from the target vocabulary
    #   at each time step.
    output_layer = Dense(
        target_vocab_size,
        kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1)
    )
    
    # Use the your decoding_layer_train() function to get the training logits.
    with tf.variable_scope("decode"):
        training_output = decoding_layer_train(
            encoder_state   = encoder_state, 
            dec_cell        = dec_cell, 
            dec_embed_input = dec_embed_input, 
            target_sequence_length = target_sequence_length, 
            max_summary_length     = max_target_sequence_length, 
            output_layer    = output_layer, 
            keep_prob       = keep_prob,
            parallel_iterations    = parallel_iterations
        ) 

    # Use your decoding_layer_infer() function to get the inference logits.
    start_of_sequence_id = target_vocab_to_int['<GO>']
    end_of_sequence_id   = target_vocab_to_int['<EOS>']
    with tf.variable_scope("decode", reuse=True):
        inference_output = decoding_layer_infer(
            encoder_state  = encoder_state, 
            dec_cell       = dec_cell, 
            #dec_embeddings = dec_embed_input, ########################uhhhh
            dec_embeddings = decoder_embedding,
            start_of_sequence_id = start_of_sequence_id, 
            end_of_sequence_id   = end_of_sequence_id, 
            max_target_sequence_length = max_target_sequence_length, 
            vocab_size   = target_vocab_size, 
            output_layer = output_layer, 
            batch_size   = batch_size, 
            keep_prob    = keep_prob,
            parallel_iterations = parallel_iterations
        )

    return (training_output, inference_output)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [41]:
def seq2seq_model(
    input_data, 
    target_data, 
    keep_prob, 
    batch_size,
    source_sequence_length, 
    target_sequence_length,
    max_target_sentence_length,
    source_vocab_size, 
    target_vocab_size,
    enc_embedding_size, 
    dec_embedding_size,
    rnn_size, 
    num_layers, 
    target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # Pass the input data through the encoder. 
    _, enc_state = encoding_layer(
        input_data, 
        rnn_size, 
        num_layers, 
        keep_prob,
        source_sequence_length,
        source_vocab_size, 
        enc_embedding_size
    )
    
    # Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(
        target_data, 
        target_vocab_to_int, 
        batch_size
    )

    # Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(
        dec_input,
        enc_state,
        target_sequence_length,
        max_target_sentence_length,
        rnn_size,
        num_layers, 
        target_vocab_to_int, 
        target_vocab_size,
        batch_size,
        keep_prob,
        dec_embedding_size                                                                       
    )       
    
    return training_decoder_output, inference_decoder_output
    #return None, None


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

### Minor Issues in 1st Submission
1. Since the vocabulary was small, I chose "50" as my encoder/decoder embedding size. However, the reviewer felt this was too small to "capture the rich semantics of words." Recommendation:  Try setting it to around 128/256 (note that powers of 2 are not only customary, but can potentially help performance).
2. rnn_size -- should be set to a power of 2, e.g., 64.

### Reviewer1's Comments on Embedding Size
> 50 as an embedding size it too small to capture the rich semantics of words. The larger your embedding size is, the more concepts your network will be able to learn, but the longer it will take to train. Try setting it to around 128/256.

> Embedding size: I've actually found smaller vectors to work better (64 and 128 - note the binary sizes, I think this helps performance by assisting theano/TF with copying chunks of data to and from GPU), likely as that's fewer parameters to learn, and too many can make learning hard for a neural network. Cross validation would help you determine a good size, try varying in magnitude (32,64,128,256) rather than more linear scales, as the relationship between these items and performance tend to be more of an exponential than a linear nature.

So... seems 128 would be the sweet spot given these two comments.  



In [42]:
# Number of Epochs
epochs = 25
# Batch Size
batch_size = 1028
# RNN Size  (how wide is the LSTM cell? i.e., how many units?)
rnn_size = 64
# Number of Layers
num_layers = 1
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.8
display_step = 100



### Build the Graph
Build the graph using the neural network you implemented.

In [43]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [44]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [45]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, 
 valid_sources_lengths, valid_targets_lengths ) = \
    next(get_batches(valid_source, valid_target, batch_size,
                     source_vocab_to_int['<PAD>'], target_vocab_to_int['<PAD>']))   
    
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch  100/134 - Train Accuracy: 0.5460, Validation Accuracy: 0.5547, Loss: 0.8292
Epoch   1 Batch  100/134 - Train Accuracy: 0.7376, Validation Accuracy: 0.7451, Loss: 0.4646
Epoch   2 Batch  100/134 - Train Accuracy: 0.8067, Validation Accuracy: 0.8072, Loss: 0.3363
Epoch   3 Batch  100/134 - Train Accuracy: 0.8402, Validation Accuracy: 0.8335, Loss: 0.2571
Epoch   4 Batch  100/134 - Train Accuracy: 0.8638, Validation Accuracy: 0.8613, Loss: 0.2118
Epoch   5 Batch  100/134 - Train Accuracy: 0.8747, Validation Accuracy: 0.8815, Loss: 0.1831
Epoch   6 Batch  100/134 - Train Accuracy: 0.8895, Validation Accuracy: 0.8819, Loss: 0.1583
Epoch   7 Batch  100/134 - Train Accuracy: 0.8968, Validation Accuracy: 0.9009, Loss: 0.1421
Epoch   8 Batch  100/134 - Train Accuracy: 0.9073, Validation Accuracy: 0.9126, Loss: 0.1327
Epoch   9 Batch  100/134 - Train Accuracy: 0.9138, Validation Accuracy: 0.9125, Loss: 0.1218
Epoch  10 Batch  100/134 - Train Accuracy: 0.9176, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

In [47]:
save_path

'checkpoints/dev'

# Checkpoint

In [48]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [49]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    sentence = sentence.lower().split()
    id_seq = [vocab_to_int[word] 
              if word in list(vocab_to_int.keys()) else vocab_to_int['<UNK>'] 
              for word in sentence]
    return id_seq


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [54]:
translate_sentence = 'this winter is like chinese snake bananas in new jeresey: difficult like a shark'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [75, 19, 210, 15, 167, 106, 207, 214, 197, 2, 125, 15, 5, 95]
  English Words: ['this', 'winter', 'is', 'like', 'chinese', 'snake', 'bananas', 'in', 'new', '<UNK>', 'difficult', 'like', 'a', 'shark']

Prediction
  Word Ids:      [179, 164, 98, 232, 250, 87, 153, 134, 1]
  French Words: cette souris est mon animal le plus aimé. <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.

### Recommended Reading
* Learn more about unit test: [The Hitchhiker's Guide to Python](http://docs.python-guide.org/en/latest/writing/tests/)
* Chris Olah: [Attention and Augmented Recurrent Neural Networks](http://distill.pub/2016/augmented-rnns/)
* Google's [tf-seq2seq](https://research.googleblog.com/2017/04/introducing-tf-seq2seq-open-source.html)
    - code on [Github](https://github.com/google/seq2seq)
* Video on [Seq2Seq](https://www.youtube.com/watch?v=G5RY_SUJih4&feature=youtu.be&t=38m14s)